### ISSUE TITLE
CMake build issues with 3.12: target already exists

### ISSUE THREAD

**Issue by GMNGeoffrey** [google/iree#365]
The GitHub cmake action uses cmake 3.12, which is having some strange issues that are not present with 3.13

Some earlier discussion in https://github.com/google/iree/pull/354

```
CMake Error at third_party/benchmark/src/CMakeLists.txt:20 (add_library):
  add_library cannot create target "benchmark" because another target with
  the same name already exists.  The existing target is a static library
  created in source directory
  "/usr/local/google/home/gcmn/git/iree-fork/third_party/llvm-project/llvm/utils/benchmark/src".
  See documentation for policy CMP0002 for more details.


CMake Error at third_party/benchmark/src/CMakeLists.txt:31 (target_link_libraries):
  Attempt to add link library "-pthread" to target "benchmark" which is not
  built in this directory.


CMake Error at third_party/benchmark/src/CMakeLists.txt:34 (target_link_libraries):
  Attempt to add link library "/usr/lib/x86_64-linux-gnu/librt.so" to target
  "benchmark" which is not built in this directory.


CMake Error at third_party/benchmark/src/CMakeLists.txt:56 (add_library):
  add_library cannot create target "benchmark_main" because another target
  with the same name already exists.  The existing target is a static library
  created in source directory
  "/usr/local/google/home/gcmn/git/iree-fork/third_party/llvm-project/llvm/utils/benchmark/src".
  See documentation for policy CMP0002 for more details.
```

**Comment by GMNGeoffrey** [comment]
The CI was last passing at https://github.com/google/iree/commit/1f260818a525f5ca712d9e72b891c1b10b86fdd2

There are some other types of failures in there, but I'm going to try and see if I can track down whether it's just a matter of the compiler build being turned on or if I can see where the failure was introduced

**Comment by GMNGeoffrey** [comment]
Yeah this is just from turning the compiler on. I checked out https://github.com/google/iree/commit/1f260818a525f5ca712d9e72b891c1b10b86fdd2 and turned the compiler on with cmake 3.12 and was able to repro the issue. Taking out the compiler again fixes it.

So one option we have here is reopening https://github.com/google/iree/pull/307. Then we don't have to boil the ocean getting the compiler building before we can have a usable CI again.

**Comment by GMNGeoffrey** [comment]
Nope, reopening #307 isn't an option. It's including the llvm-project directory that triggers it, regardless of whether compiler is on or off

**Comment by GMNGeoffrey** [comment]
So remaining options:
1. Figure out what the issue is and make this build with 3.12
2. Get the CI to use 3.13

I've tried to do (2) in the past and it ended up being way more of a pain than I thought was reasonable. I ran into some weirdo issues. It would be nice to continue supporting 3.12 also. I don't think it's *that* old. It's the default on a few different platforms I've run into, at least.

(1) is also not super appealing, but I'll start with that

**Comment by GMNGeoffrey** [comment]
So these are really the same library getting included in two different ways. LLVM just has a copy of google/benchmark inside of it. There's nothing in the 3.13 release notes about add_library behavior changing. There is [something about target_sources() now treating sources as relative](https://cmake.org/cmake/help/v3.13/release/3.13.html#other-changes), but I don't think that applies. AFAICT, since CMP0002, these names have to be unique, but both cmake 3.12 and 3.13 [use the OLD behavior by default](https://cmake.org/cmake/help/v3.12/policy/CMP0002.html)

**Comment by GMNGeoffrey** [comment]
So the other option we have is turning off IREE_BUILD_TESTS (and move the googletest dependency under that, which I think it probably should be anyway). I'm still seeing failures, but they look more like the "Ben deleted this directory" sort

**Comment by GMNGeoffrey** [comment]
Ok my current theory is that this has to do with the change in behavior for target_link_libraries and target_sources, which we use in iree_cc_library. There's a blog post about this new shiny behavior, which I skimmed. It's got some info on how to get 3.12 compatibility: https://crascit.com/2016/01/31/enhanced-source-file-handling-with-target_sources/

**Comment by GMNGeoffrey** [comment]
Ok I'm going back to option 2. Let's see if I can get cmake 3.13 to install.

**Comment by GMNGeoffrey** [comment]
Step 1, uninstall cmake 3.12, is not working as expected :-D https://github.com/google/iree/pull/367/checks?check_run_id=371845578

I'll finish this tomorrow

**Comment by GMNGeoffrey** [comment]
@marbre has figured out how to make this work with cmake 3.12, so we can stick with that.

**Comment by GMNGeoffrey** [comment]
Fixed via #368

**Comment by GMNGeoffrey** [comment]
Fixed (#368)
